File: python/tests/test_taxonomy.py
"""The error taxonomy has an enforcer (contract v1, grammar.md section 6).

Two things are asserted here, and neither had ever been asserted before:

1. **Doc truth.** ``config/grammar.md`` section 6 and
   ``tests/error_taxonomy.csv`` must agree row for row. Nothing in any suite had
   ever *read* ``grammar.md`` -- every reference to it was a prose citation --
   which is how its taxonomy table came to document an accidental Stata rc 601
   leak as though it were the design, and to claim a ``source_stage`` vocabulary
   that was "byte-identical across languages" immediately before listing three
   per-language exceptions.

2. **Code truth, for Python's column.** Each row's exception name must exist in
   :mod:`datalib.errors` and must actually be raised by a real trigger. Asserting
   the name exists is not enough: a name can survive a rename while nothing
   raises it any more.

The R and Stata columns are asserted by their own suites
(``test-taxonomy.R``, ``run_conformance.do`` Part 8), because only that language
can trigger its own errors.
"""

from __future__ import annotations

import getpass
import re
from pathlib import Path

import pytest

import datalib as dl
from datalib import errors as E

from conftest import read_cases

ROWS = read_cases("error_taxonomy.csv")


# --------------------------------------------------------------------- doc truth


def _grammar_table(repo_root: Path) -> list[dict[str, str]]:
    """Parse the section 6 taxonomy table out of config/grammar.md."""
    text = (repo_root / "config" / "grammar.md").read_text(encoding="utf-8")
    header = "| contract error | Stata rc | R condition classes | Python exception |"
    assert header in text, (
        "config/grammar.md section 6 no longer has the expected taxonomy table "
        "header; the doc-truth test cannot find the table it must check"
    )
    out: list[dict[str, str]] = []
    for line in text.split(header, 1)[1].splitlines():
        raw = line.strip()
        if not raw.startswith("|"):
            if out:
                break                            # table ended
            continue                             # not into the rows yet
        cells = [c.strip() for c in raw.strip("|").split("|")]
        # the |---|---| separator row
        if all(set(c) <= {"-", ":"} for c in cells):
            continue
        assert len(cells) == 4, f"malformed taxonomy row in grammar.md: {raw!r}"
        # Cells are prose: a value may or may not be backticked, and a cell may
        # hold several space-separated tokens (`198` `601`). Normalise by dropping
        # backticks and collapsing whitespace, so the CSV is compared on values
        # rather than on markdown decoration.
        norm = [re.sub(r"\s+", " ", c.replace("`", "")).strip() for c in cells]
        out.append(
            {
                "error_name": norm[0],
                "stata_rc": norm[1],
                "r_classes": norm[2],
                "python_exc": norm[3],
            }
        )
    return out


def test_grammar_table_matches_the_csv(repo_root: Path) -> None:
    """grammar.md section 6 is the prose; the CSV is the machine-readable form."""
    assert _grammar_table(repo_root) == [dict(r) for r in ROWS]


def test_taxonomy_csv_covers_every_contract_error() -> None:
    assert [r["error_name"] for r in ROWS] == [
        "input_invalid",
        "not_found",
        "config_file_missing",
        "user_block_missing",
        "root_unset",
    ]


# -------------------------------------------------------------------- code truth


@pytest.mark.parametrize("row", ROWS, ids=[r["error_name"] for r in ROWS])
def test_python_exception_name_exists(row) -> None:
    exc = getattr(E, row["python_exc"], None)
    assert exc is not None, (
        f"{row['error_name']}: grammar.md names {row['python_exc']}, "
        f"which does not exist in datalib.errors"
    )
    assert issubclass(exc, E.DatalibError)


def _trigger(error_name: str, fixture_root, tmp_path, monkeypatch):
    """A real call that must raise the row's exception."""
    if error_name == "input_invalid":
        return lambda: dl.datalib_vintages(
            "ZWE", "notayear", "MICS", root=fixture_root
        )
    if error_name == "not_found":
        # CountryNotFound is a NotFoundError subclass, so the documented parent
        # is what a portable caller catches.
        return lambda: dl.datalib_resolve("XXX", 2019, "MICS", root=fixture_root)
    if error_name == "config_file_missing":
        return lambda: dl.datalib_config(config=tmp_path / "no_such_file.yml")
    if error_name == "user_block_missing":
        cfg = tmp_path / "user_config.yml"
        cfg.write_text('someoneelse:\n  datalib: "Z:/x"\n', encoding="utf-8")
        return lambda: dl.datalib_config(user="nobody", config=cfg)
    if error_name == "root_unset":
        # A block that exists for this user but carries no datalib: key -- the
        # only path to DatalibRootNotSet, since a missing file raises
        # ConfigFileNotFound instead.
        cfg = tmp_path / "user_config.yml"
        cfg.write_text(f"{getpass.getuser()}:\n  githubFolder: \"C:/GitHub\"\n",
                       encoding="utf-8")
        monkeypatch.delenv("DATALIB_ROOT", raising=False)
        monkeypatch.setenv("DATALIB_CONFIG", str(cfg))
        return dl.datalib_root
    raise AssertionError(f"no trigger defined for {error_name!r}")


@pytest.mark.parametrize("row", ROWS, ids=[r["error_name"] for r in ROWS])
def test_documented_exception_is_the_one_raised(
    row, fixture_root, tmp_path, monkeypatch
) -> None:
    exc = getattr(E, row["python_exc"])
    with pytest.raises(exc):
        _trigger(row["error_name"], fixture_root, tmp_path, monkeypatch)()

File: r/tests/testthat/test-taxonomy.R
# The error taxonomy has an enforcer, for R's column (grammar.md section 6).
#
# tests/error_taxonomy.csv is the machine-readable mapping; this file asserts
# R's share of it. Two things matter and neither was asserted before v0.9.7:
#
#   1. Every class the CSV names for a row is actually on the condition R raises
#      -- asserting the constructor exists is not enough, because a name can
#      survive a rename while nothing raises it any more.
#   2. The class vector is ordered specific-then-general, so a handler written
#      against datalib_error_input before the subclasses existed keeps firing.
#      That property is what makes adding an R subclass a non-breaking change,
#      and it is the reason the subclasses could land outside a major release.
#
# Python's column is asserted in test_taxonomy.py (which also checks grammar.md
# against the CSV); Stata's in run_conformance.do Part 8. Only the language
# itself can trigger its own errors.

rows <- read.csv(file.path(datalib_test_repo_tests(), "error_taxonomy.csv"),
                 colClasses = "character")
fixture <- datalib_test_fixture_root()

# One real trigger per contract error. Each returns a zero-argument function.
dl_trigger <- function(error_name) {
  switch(
    error_name,
    input_invalid = function() {
      datalib_vintages("ZWE", "notayear", "MICS", root = fixture)
    },
    not_found = function() {
      datalib_surveys("XXX", root = fixture)
    },
    config_file_missing = function() {
      datalib_config(config = file.path(tempdir(), "dl_no_such_config.yml"))
    },
    user_block_missing = function() {
      d <- file.path(tempdir(), "dl_tax_block")
      dir.create(d, recursive = TRUE, showWarnings = FALSE)
      f <- file.path(d, "user_config.yml")
      writeLines(c("someoneelse:", '  datalib: "Z:/x"'), f)
      datalib_config(user = "nobody", config = f)
    },
    root_unset = function() {
      # An empty config dir with no env var and no option: nothing can resolve.
      d <- file.path(tempdir(), "dl_tax_emptycfg")
      dir.create(d, recursive = TRUE, showWarnings = FALSE)
      withr::with_envvar(
        c(DATALIB_ROOT = NA, DATALIB_CONFIG = NA, DATALIB_CONFIG_DIR = NA),
        withr::with_options(
          list(datalib.root = NULL),
          datalib_root(configdir = d)
        )
      )
    },
    stop("no trigger defined for ", error_name)
  )
}

for (i in seq_len(nrow(rows))) {
  row <- rows[i, ]
  classes <- strsplit(trimws(row$r_classes), "[[:space:]]+")[[1]]

  label <- paste0("taxonomy ", row$error_name, ": documented classes are raised")
  test_that(label, {
    cond <- tryCatch(dl_trigger(row$error_name)(),
                     condition = function(c) c)
    expect_s3_class(cond, "condition")
    expect_s3_class(cond, "datalib_error")
    for (cl in classes) {
      expect_true(inherits(cond, cl),
                  info = paste0(row$error_name, " must carry class ", cl,
                                "; got: ", paste(class(cond), collapse = " ")))
    }
    # specific-then-general: the documented order is the class-vector order
    expect_identical(class(cond)[seq_along(classes)], classes)
  })
}

test_that("a subclassed error is still caught by a pre-v0.9.7 handler", {
  # The backward-compatibility property, asserted directly rather than assumed:
  # code written against datalib_error_input before datalib_error_not_found
  # existed must keep working.
  caught <- tryCatch(
    datalib_surveys("XXX", root = fixture),
    datalib_error_input = function(e) "general handler fired",
    error = function(e) "fell through"
  )
  expect_identical(caught, "general handler fired")
})

test_that("the taxonomy covers every contract error, in order", {
  expect_identical(rows$error_name,
                   c("input_invalid", "not_found", "config_file_missing",
                     "user_block_missing", "root_unset"))
})
